### PR TITLE
[api] simplify user state code

### PR DESCRIPTION
**Pull request by adrianschroeter**
* drop not anymore used user states
* drop User::STATES array and use an enum instead
* drop some over designed code

### CODE REVIEW

**Comment by hennevogel** [comment]
http://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html

**Comment by adrianschroeter** [comment]
we can cleanup more here, but as a start....

I suppose henne will recommend a state validation :)

**Comment by adrianschroeter** [comment]
@hennevogel wasn't there the problem that the activerecord enums could be used together with mysql enums?

File: src/api/app/models/owner.rb [hunk 1]
```diff
@@ -153,18 +153,18 @@ def self.find_containers_without_definition(rootproject, devel = true, filter =
     end
 
     # find all groups which have an active user
-    maintained_groups = Group.joins(:groups_users).joins(:users).where("users.state = ?", User::STATES['confirmed']).to_a
+    maintained_groups = Group.joins(:groups_users).joins(:users).where("users.state = ?", 'confirmed').to_a
```
Comment (mdeniz) [inline src/api/app/models/owner.rb]: Better this way, isn't it?:
```ruby
maintained_groups = Group.joins(:groups_users).joins(:users).where("users.state = 'confirmed'").to_a
```

File: src/api/app/models/owner.rb [hunk 2]
```diff
@@ -153,18 +153,18 @@ def self.find_containers_without_definition(rootproject, devel = true, filter =
...
 
     # fast find packages with defintions
     # relationship in package object by user
     defined_packages = Package.where(project_id: projects).joins(:relationships => :user).where(["relationships.role_id IN (?) AND users.state = ?",
-                                                                                                 roles, User::STATES['confirmed']]).pluck(:name)
+                                                                                                 roles, 'confirmed']).pluck(:name)
```
Comment (mdeniz) [inline src/api/app/models/owner.rb]: Maybe same here, something like:
```ruby
defined_packages = Package.where(project_id: projects).joins(:relationships => :user).where("relationships.role_id IN (?) AND users.state = 'confirmed'", roles).pluck(:name)
```

File: src/api/app/models/owner.rb [hunk 3]
```diff
@@ -153,18 +153,18 @@ def self.find_containers_without_definition(rootproject, devel = true, filter =
...
     # relationship in package object by group
     defined_packages += Package.where(project_id: projects).joins(:relationships).where(["relationships.role_id IN (?) AND group_id IN (?)",
                                                                                          roles, maintained_groups]).pluck(:name)
     # relationship in project object by user
     Project.joins(:relationships => :user).where("projects.id in (?) AND role_id in (?) AND users.state = ?",
-                                                 projects, roles, User::STATES['confirmed']).each do |prj|
+                                                 projects, roles, 'confirmed').each do |prj|
```
Comment (mdeniz) [inline src/api/app/models/owner.rb]: Same here, put 'confirmed' inside the string directly

File: src/api/app/models/unregistered_user.rb [hunk 4]
```diff
@@ -41,19 +41,12 @@ def self.can_register?
...
-    return state
-  end
-
   def self.register(opts)
     can_register?
 
     opts[:note] = nil unless User.current and User.current.is_admin?
-    state = get_state
+    state = 'unconfirmed'
+    state = 'confirmed' if ::Configuration.registration == 'allow'
```
Comment (mdeniz) [inline src/api/app/models/unregistered_user.rb]: Better this way:
```ruby
state = ::Configuration.registration == 'allow' ? 'confirmed' : 'unconfirmed'
```

File: src/api/app/models/user.rb [hunk 5]
```diff
@@ -174,8 +165,8 @@ def before_create
 
   # the default state of a user based on the api configuration
   def self.default_user_state
-    return STATES['unconfirmed'] if ::Configuration.registration == "confirmation"
-    STATES['confirmed']
+    return 'unconfirmed' if ::Configuration.registration == "confirmation"
+    'confirmed'
```
Comment (mdeniz) [inline src/api/app/models/user.rb]: Better like this:
```ruby
def self.default_user_state
  ::Configuration.registration == 'confirmation' ? 'unconfirmed' : 'confirmed'
end
```

File: src/api/app/models/owner.rb (hunk 3 above)
Comment (adrianschroeter) [inline src/api/app/models/owner.rb]: On Freitag, 10. Juni 2016, 08:19:10 CEST wrote Mois�s D�niz Alem�n:
> >      # relationship in package object by group
> >      defined_packages += Package.where(project_id: projects).joins(:relationships).where(["relationships.role_id IN (?) AND group_id IN (?)",
> >                                                                                           roles, maintained_groups]).pluck(:name)
> >      # relationship in project object by user
> >      Project.joins(:relationships => :user).where("projects.id in (?) AND role_id in (?) AND users.state = ?",
> > -                                                 projects, roles, User::STATES['confirmed']).each do |prj|
> > +                                                 projects, roles, 'confirmed').each do |prj|
> 
> Same here, put 'confirmed' inside the string directly

right, all three lines changed

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imend�rffer, Jane Smithard, Graham Norton, HRB 21284 (AG N�rnberg) 
 
Maxfeldstra�e 5                         
90409 N�rnberg 
Germany

File: src/api/app/models/user.rb [hunk 6]
```diff
@@ -338,15 +324,15 @@ def self.get_default_admin
...
+                                                      state: 'locked',
                                                       password: "123456",
                                                       password_confirmation: "123456").find_or_create_by(login: nobody_login)
     Thread.current[:nobody_user]
   end
 
   def self.find_by_login!(login)
     user = find_by_login(login)
-    if user.nil? or user.state == STATES['deleted']
+    if user.nil? or user.state == 'deleted'
```
Comment (mdeniz) [inline src/api/app/models/user.rb]: Don't use `or`, use `||`

**Comment by mdeniz** [comment]
Beside the nitpicks it LGTM

**Comment by bgeuken** [comment]
LGTM

+1 for [rails's enums](https://github.com/openSUSE/open-build-service/issues/1626) in the long run

File: src/api/test/unit/user_test.rb [hunk 7]
```diff
@@ -102,10 +102,6 @@ def test_render_axml
     assert_no_xml_tag axml, :tag => :globalrole, :content => 'Admin'
   end
 
-  def test_states
-    assert_not_nil User::STATES
```
Comment (bgeuken) [inline src/api/test/unit/user_test.rb]: That's a funny one^^

**Pull request merged**